salt/utils/requisite.py

```diff
@@ -0,0 +1,457 @@
+"""
+The classes and functions in this module are related to requisite
+handling and ordering of chunks by the State compiler
+"""
+
+from __future__ import annotations
+
+import fnmatch
+import logging
+from collections import defaultdict
+from collections.abc import Generator, Iterable, Sequence
+from enum import Enum, auto
+from typing import Any
+
+import networkx as nx  # pylint: disable=3rd-party-module-not-gated
+
+log = logging.getLogger(__name__)
+
+# See https://docs.saltproject.io/en/latest/ref/states/layers.html for details on the naming
+LowChunk = dict[str, Any]
+
+
+def _gen_tag(low: LowChunk) -> str:
+    """
+    Generate the a unique identifer tag string from the low data structure
+    """
+    return "{0[state]}_|-{0[__id__]}_|-{0[name]}_|-{0[fun]}".format(low)
+
+
+def trim_req(req: dict[str, Any]) -> dict[str, Any]:
+    """
+    Trim any function off of a requisite reference
+    """
+    reqfirst, valfirst = next(iter(req.items()))
+    if "." in reqfirst:
+        return {reqfirst.split(".", maxsplit=1)[0]: valfirst}
+    return req
+
+
+class RequisiteType(str, Enum):
+    """Types of direct requisites"""
+
```

Comment:
This crashes on Python 3.9, which is supported by Salt (https://github.com/python/mypy/issues/7591#issuecomment-1376636258 / https://github.com/astral-sh/ruff/issues/2346). It might be the cause of the SSH test failures on RHEL derivatives.

Reply:
Ah thanks for the info, I was not aware of this issue on python 3.9.

Reply:
Updated to use staticmethod_hack. I still need to figure out a solution to the importlib_metadata issue though.

Reply:
I don't see anything wrong with proper one I proposed below, but I'm also not sure why it's not done like that currently.

You could check if including it fixes the rest of the integration tests and/or breaks anything else.

Reply:
Thanks for you help.

Reply:
You're welcome, hope this can land in 3008. (Thanks for creating this btw, I'm sure it was a major effort!)

Since you went pretty deep into the state compiler - if you have the capacity at some point - could you comment on https://github.com/saltstack/salt/issues/59959#issuecomment-2053688186? I would love to fix that issue (and possibly https://github.com/saltstack/salt/issues/49301) in the next release as well (based on this branch since its current state is just unwieldy).